2.evaluate_data/utils/DeterministicSampling.py

```diff
@@ -57,26 +46,18 @@ def sample_plate_deterministically(self, _sample_strategy: str = "well_sampling"
             ) * self._divisor
             return self._platedf.loc[self._platedf["Metadata_farmhash"] < mod_cutoff]
 
-        grouped_platedf = self._platedf.groupby(self._group_cols)
-
-        num_groups = len(grouped_platedf.groups)
-
-        # Samples approximately this number, per group, from the plate dataframe self._platedf
-        samples_per_group = self._samples_per_plate // num_groups
+        grouped = self._platedf.groupby(self._group_cols)
 
-        if samples_per_group < 1:
-            warnings.warn(
-                "The number of samples per group must be at least 1. Changing this to 1."
-            )
-            samples_per_group = 1
+        num_groups = len(grouped)
+        samples_per_group = max(1, self._samples_per_plate // num_groups)
 
-        sampled_platedf = []
-
-        for _, group_df in grouped_platedf:
-            num_samples_group = group_df.shape[0]
-            mod_cutoff = (samples_per_group / num_samples_group) * self._divisor
-            sampled_platedf.append(
-                group_df.loc[group_df["Metadata_farmhash"] < mod_cutoff]
-            )
-
-        return pd.concat(sampled_platedf, axis=0)
+        return pd.concat(
+            (
+                group_df.loc[
+                    group_df["Metadata_farmhash"]
+                    < (samples_per_group / len(group_df)) * self._divisor
+                ]
+                for _, group_df in grouped
+            ),
+            ignore_index=True,
+        )
```

Comment:
This looks like complex logic; consider adding a comment or two about how the grouping works and what the product of this comprehension will be.